### PR TITLE
[SourceKit] SourceKit support for opaque result types

### PR DESCRIPTION
**Pull request by rintaro**
Initial implementation for SourceKit support for "opaque result types"

* libSyntax: Grammer and parser change
* Syntax highlighting, Cursor Info, Indexing
* Code completion for `some` keyword
* Conformance completion using opaque result types

rdar://problem/49997412

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -3402,7 +3409,27 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
...
+        switch (Options.OpaqueReturnTypePrinting) {
+        case PrintOptions::OpaqueReturnTypePrintingMode::StableReference:
+        case PrintOptions::OpaqueReturnTypePrintingMode::Description:
+          isSimple = true;
+          break;
+        case PrintOptions::OpaqueReturnTypePrintingMode::WithOpaqueKeyword:
+          isSimple = false;
+          break;
+        case PrintOptions::OpaqueReturnTypePrintingMode::WithoutOpaqueKeyword: {
+          isSimple = opaqueTy->getConformsTo().size() < 2;
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp]: Can an opaque type have a superclass and one conformance? CC @jckarter

File: lib/AST/ASTPrinter.cpp [hunk 2]
```diff
@@ -4175,16 +4202,28 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
...
-    OpaqueTypeDecl *decl = T->getDecl();
-    
-    if (Options.PrintStableReferencesToOpaqueReturnTypes) {
+    switch (Options.OpaqueReturnTypePrinting) {
+    case PrintOptions::OpaqueReturnTypePrintingMode::WithOpaqueKeyword:
+      Printer << "some ";
+      LLVM_FALLTHROUGH;
+    case PrintOptions::OpaqueReturnTypePrintingMode::WithoutOpaqueKeyword: {
+      SmallVector<Type, 2> types;
+      for (auto proto : T->getConformsTo())
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp]: What about T->getSuperclass()?

Maybe this should be generalized to get all the constraints from `T->getBoundSignature()` instead?

File: lib/AST/ASTPrinter.cpp [hunk 3]
```diff
@@ -4175,16 +4202,28 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
...
+        types.push_back(proto->TypeDecl::getDeclaredInterfaceType());
+
+      // Create and visit temporary ProtocolCompositionType.
+      auto composition =
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp]: Is there a more direct way to turn an opaque type into an existential, @jckarter ?

File: lib/IDE/CodeCompletion.cpp [hunk 4]
```diff
@@ -4025,27 +4038,82 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
...
+    // Try substitution to see if the associated type is resolved to concrete
+    // type.
+    auto substMap = currTy->getMemberSubstitutionMap(
+        CurrDeclContext->getParentModule(), VD);
+    ResultT = ResultT.subst(substMap, SubstFlags::UseErrorType);
+    if (!ResultT || !ResultT->is<DependentMemberType>())
+      // If resolved print it.
+      return nullptr;
+
+    return assocTyD->getInherited()[0];
```
Comment (slavapestov) [inline lib/IDE/CodeCompletion.cpp]: What if getInherited() has multiple entries?

File: lib/IDE/CodeCompletion.cpp [hunk 5]
```diff
@@ -4025,27 +4038,82 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
...
+    if (auto *FD = dyn_cast<FuncDecl>(VD))
+      ResultT = FD->getResultInterfaceType();
+    else if (auto *SD = dyn_cast<SubscriptDecl>(VD))
+      ResultT = SD->getElementInterfaceType();
+    else if (auto *VarD = dyn_cast<VarDecl>(VD))
+      ResultT = VarD->getInterfaceType();
+    else
+      return nullptr;
+
+    if (!ResultT->is<DependentMemberType>())
```
Comment (slavapestov) [inline lib/IDE/CodeCompletion.cpp]: I'm not sure what this function is doing. Are you printing protocol requirements and extension methods that return associated types as returning an opaque type?

Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]: Yes, for example:
```swift
protocol P {
  associatedtype Assoc: Collection
  func foo() -> Assoc
}
protocol C: P {
  func // Complete here.
}
```
We wan't to complete `foo()` as `func foo() -> some Collection { <#code#> }`

File: lib/IDE/CodeCompletion.cpp (hunk 4 above)
Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]:
```swift
  associatedtype Aaaa: Collection, Comparable
```
Thanks! I didn't know this is accepted by the compiler.

File: lib/AST/ASTPrinter.cpp (hunk 2 above)
Comment (rintaro) [reply lib/AST/ASTPrinter.cpp]: Will fix.

File: lib/IDE/CodeCompletion.cpp (hunk 4 above)
Comment (slavapestov) [reply lib/IDE/CodeCompletion.cpp]: In fact the inheritance clause doesn’t contain all the information you need. I would suggest asking the protocol’s generic signature instead.

Eg, `protoDecl->getGenericSignature()->getConformsTo(assocDecl->getDeclaredInterfaceType())`

You should also check for a superclass and layout constraint, so you can print ‘some SomeClass’ or ‘some AnyObject & P’, etc.

File: lib/AST/ASTPrinter.cpp (hunk 1 above)
Comment (jckarter) [reply lib/AST/ASTPrinter.cpp]: Yes, an opaque type can have any combination of constraints.

File: lib/AST/ASTPrinter.cpp (hunk 3 above)
Comment (jckarter) [reply lib/AST/ASTPrinter.cpp]: I added `ArchetypeType::getExistentialType()` in https://github.com/apple/swift/pull/24166.

Comment (rintaro) [reply lib/AST/ASTPrinter.cpp]: @jckarter Thanks! I'll modify ASTPrinter to use that too.

**Pull request merged**